spinup_bis/algos/tf2/bsp/sop.py

```diff
@@ -0,0 +1,512 @@
+            model. Setting the value to None turns off the saving.
+
+
+        seed (int): Seed for random number generators.
+    """
+    pwd = os.getcwd()  # pylint: disable=possibly-unused-variable
+    config = locals()
+    logger = logx.EpochLogger(**(logger_kwargs or {}))
+    logger.save_config(config)
+
+
+    random.seed(seed)
+    tf.random.set_seed(seed)
+    np.random.seed(seed)
+
+    env, test_env = env_fn(), env_fn()
+    obs_dim = env.observation_space.shape[0]
+    act_dim = env.action_space.shape[0]
+    # This implementation assumes all dimensions share the same bound!
+    assert np.all(env.action_space.high == env.action_space.high[0])
+
+    # Share information about action space with policy architecture
+    ac_kwargs = ac_kwargs or {}
+    ac_kwargs['observation_space'] = env.observation_space
+    ac_kwargs['action_space'] = env.action_space
+    ac_kwargs['act_noise'] = act_noise
+    ac_kwargs['ac_number'] = ac_number
+
+    # Network
+    ac_factory = actor_critic(**ac_kwargs)
+
+    actor = ac_factory.make_actor()
+    actor.build(input_shape=(None, obs_dim))
+
+    critic1 = ac_factory.make_critic()
+    critic1.build(input_shape=(None, obs_dim + act_dim))
+    critic2 = ac_factory.make_critic()
+    critic2.build(input_shape=(None, obs_dim + act_dim))
+
+    critic_variables = critic1.trainable_variables + critic2.trainable_variables
+
+    # Target networks
+    target_critic1 = ac_factory.make_critic()
+    target_critic1.build(input_shape=(None, obs_dim + act_dim))
+    target_critic2 = ac_factory.make_critic()
+    target_critic2.build(input_shape=(None, obs_dim + act_dim))
+
+    # Copy weights
+    target_critic1.set_weights(critic1.get_weights())
+    target_critic2.set_weights(critic2.get_weights())
+
+    # Experience buffer
+    replay_buffer = ReplayBuffer(obs_dim=obs_dim,
+                                 act_dim=act_dim,
+                                 size=replay_size,
+                                 ac_number=ac_number,
+                                 max_ep_len=max_ep_len,
+                                 init_ere_coeff=init_ere_coeff)
+
+    # Separate train ops for pi, q
+    optimizer = tf.keras.optimizers.Adam(learning_rate=lr)
+
+    @tf.function
+    def vote_evaluation_policy(obs):
+        obs_actor = tf.broadcast_to(obs, [ac_number, 1, *obs.shape])
+        obs_critic = tf.broadcast_to(obs, [ac_number, ac_number, *obs.shape])
+
+        mu, _ = actor(obs_actor)
+        # One action per batch.
+        act = tf.reshape(mu, [1, ac_number, *mu.shape[2:]])
+        # The same action for each component.
+        act = tf.broadcast_to(act, [ac_number, ac_number, *mu.shape[2:]])
+        # Evaluate each action by all components.
+        qs = critic1([obs_critic, act])
+        # Average over ensemble.
+        qs = tf.reduce_mean(qs, axis=0)
+
+        return mu[tf.math.argmax(qs)][0]
+
+    @tf.function
+    def mean_evaluation_policy(obs):
```

Comment:
Remake into eval. policy without ensemble averaging (remove ensemble from this code).

Reply:
Possibly you don't need broadcast (see SAC, this will be similar).

Reply:
done